testcluster: disallow sharing of Settings

RaduBerinde opened the pull request:
If a test sets `TestClusterArgs.ServerArgs.Settings`, the `Settings`
object will be shared between all nodes. This can lead to very hard to
root-cause failures, especially around version changes (including
during node initialization).

As part of this change, we also correct the unnecessary embedding of
`*Settings` and `RaftSettings` into this struct.

Fixes: #112395
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/112396)
<!-- Reviewable:end -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 7 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112396#-:-Ngsp7ntBw3ojieeTMpn:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DarrylWong, @dt, @RaduBerinde, and @srosenberg)

___
*[`pkg/base/test_server_args.go` line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112396#-Ngso0xW2X3RrD-U4gfh:-Ngso0xW2X3RrD-U4gfi:b-isnwbx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ede74a66aee69dab60e537d69364fe6c8e1d667f/pkg/base/test_server_args.go#L38)):*
> ```Go
> // each node. If any fields are added which are not safe to share (like
> // Settings), a check that the field is unset should be added in
> // StartTestCluster (when the cluster has more than 1 node).
> ```

nit: the check is in NewTestCluster which is called by StartTestCluster, not sure if you consider that a discrepancy


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
testcluster: disallow sharing of Settings

If a test sets `TestClusterArgs.ServerArgs.Settings`, the `Settings`
object will be shared between all nodes. This can lead to very hard to
root-cause failures, especially around version changes (including
during node initialization).

As part of this change, we also correct the unnecessary embedding of
`*Settings` and `RaftSettings` into this struct.

Fixes: #112395
Release note: None

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112396#-:-NgtgFr_3hVA-mzlfkod:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DarrylWong, @dt, @mgartner, and @srosenberg)

___
*[`pkg/base/test_server_args.go` line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112396#-Ngso0xW2X3RrD-U4gfh:-NgtgHDg-NHG5KIUZF-G:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ede74a66aee69dab60e537d69364fe6c8e1d667f/pkg/base/test_server_args.go#L38)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: the check is in NewTestCluster which is called by StartTestCluster, not sure if you consider that a discrepancy
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Oh, wow, there are a ton of tests that fail now :/

RaduBerinde commented:
Closing for now, will open a new one when I have something.
